frame/election-provider-multi-phase/src/signed.rs

```diff
@@ -0,0 +1,662 @@
+					);
+					found_solution = true;
+
+					weight = weight
+						.saturating_add(T::WeightInfo::finalize_signed_phase_accept_solution());
+					break;
+				}
+				Err(_) => {
+					Self::finalize_signed_phase_reject_solution(&who, deposit);
+					weight = weight
+						.saturating_add(T::WeightInfo::finalize_signed_phase_reject_solution());
+				}
+			}
+		}
+
+		// Any unprocessed solution is pointless to even consider. Feasible or malicious,
+		// they didn't end up being used. Unreserve the bonds.
+		for not_processed in all_submissions {
+			let SignedSubmission { who, deposit, .. } = not_processed;
+			let _remaining = T::Currency::unreserve(&who, deposit);
+			weight = weight.saturating_add(T::DbWeight::get().writes(1));
+			debug_assert!(_remaining.is_zero());
+		};
+
+		(found_solution, weight)
+	}
+
+	/// Helper function for the case where a solution is accepted in the signed phase.
+	///
+	/// Extracted to facilitate with weight calculation.
+	///
+	/// Infallible
+	pub fn finalize_signed_phase_accept_solution(
+		ready_solution: ReadySolution<T::AccountId>,
+		who: &T::AccountId,
+		deposit: BalanceOf<T>,
+		reward: BalanceOf<T>,
+	) {
+		// write this ready solution.
+		<QueuedSolution<T>>::put(ready_solution);
+
+		// unreserve deposit.
+		let _remaining = T::Currency::unreserve(who, deposit);
+		debug_assert!(_remaining.is_zero());
+
+		// Reward.
+		let positive_imbalance = T::Currency::deposit_creating(who, reward);
+		T::RewardHandler::on_unbalanced(positive_imbalance);
+	}
+
+	/// Helper function for the case where a solution is accepted in the rejected phase.
+	///
+	/// Extracted to facilitate with weight calculation.
+	///
+	/// Infallible
+	pub fn finalize_signed_phase_reject_solution(who: &T::AccountId, deposit: BalanceOf<T>) {
+		let (negative_imbalance, _remaining) = T::Currency::slash_reserved(who, deposit);
+		debug_assert!(_remaining.is_zero());
+		T::SlashHandler::on_unbalanced(negative_imbalance);
+	}
+
+	/// Insert a solution into the queue while maintaining an ordering by solution quality.
+	///
+	/// Solutions are ordered in reverse: strong solutions have low indices.
+	///
+	/// If insertion was successful, `Some(index)` is returned where index is the
+	/// index of the newly inserted item.
+	///
+	/// Note: this function maintains the invariant that `queue.len() <= T::SignedMaxSubmissions`.
+	/// In the event that insertion would violate that invariant, the weakest element is dropped.
+	///
+	/// Invariant: The returned index is always a valid index in `queue` and can safely be used to
+	/// inspect the newly inserted element.
+	pub fn insert_submission(
+		who: &T::AccountId,
+		queue: &mut Vec<SignedSubmission<T::AccountId, BalanceOf<T>, CompactOf<T>>>,
+		solution: RawSolution<CompactOf<T>>,
+		size: SolutionOrSnapshotSize,
+	) -> Option<usize> {
+		// Let's ensure that our input is valid.
+		let max_submissions = T::SignedMaxSubmissions::get();
+		debug_assert!(queue.len() as u32 <= max_submissions);
```

Comment:
lets just start using `BoundedVec`?

Reply:
Are you ok putting off that work to https://github.com/paritytech/substrate/pull/8933? When the bulk of this PR was written, none of the bounded data structures were actually available yet, but I agree that they could make things substantially simpler.

Reply:
regardless of `BoundedWhat`, this queue is a great candidate to be _strictly_ bounded. as noted, I think vec would be fine, but we can experiment with btreeset as well, if there's good reason for it.